File: algorithms/[1442]形成两个异或相等数组的三元组数目/solution.py
# topics = ["位运算"]

from typing import List


class Solution:
    def countTriplets(self, arr: List[int]) -> int:
        """
        遍历 + 异或前缀和
        time O(n^3), space O(n), n 为 arr 长度
        """
        n = len(arr)
        res = 0

        # 计算异或前缀和，preXor[i] 表示前 i 个数的异或前缀和
        preXor = [0]

        for num in arr:
            preXor.append(preXor[-1] ^ num)

        for i in range(n):
            for j in range(i + 1, n):
                for k in range(j, n):
                    # 易证 a == b 等价于 preXor[k + 1] == preXor[i]
                    if preXor[k + 1] == preXor[i]:
                        res += 1

        return res

    def countTriplets2(self, arr: List[int]) -> int:
        """
        遍历（优化） + 异或前缀和
        time O(n^2), space O(n), n 为 arr 长度
        """
        n = len(arr)
        res = 0

        # 计算异或前缀和，preXor[i] 表示前 i 个数的异或前缀和
        preXor = [0]

        for num in arr:
            preXor.append(preXor[-1] ^ num)

        for i in range(n):
            for k in range(i + 1, n):
                # 易证 a == b 等价于 preXor[k + 1] == preXor[i]
                if preXor[k + 1] == preXor[i]:
                    # [i + 1, k] 范围内的任意 j 都符合要求
                    res += k - i

        return res

File: algorithms/[1442]形成两个异或相等数组的三元组数目/solution.ts
// topics = ["位运算"]

/**
 * Bit Manipulation & Hash Table
 * time O(n), space O(n), n 为 arr 长度
 */
function countTriplets(arr: number[]): number {
  const n = arr.length;
  const s = [0];
  for (const num of arr) {
    s.push(s[s.length - 1] ^ num);
  }

  // 哈希表，空间换时间
  const cnt = new Map<number, number>();
  const total = new Map<number, number>();

  let res = 0;
  for (let k = 0; k < n; k += 1) {
    if (cnt.has(s[k + 1])) {
      // 参考 https://leetcode-cn.com/problems/count-triplets-that-can-form-two-arrays-of-equal-xor/solution/xing-cheng-liang-ge-yi-huo-xiang-deng-sh-jud0/
      res += cnt.get(s[k + 1])! * k - total.get(s[k + 1])!;
    }

    // 统计 S[k] 的出现次数
    cnt.set(s[k], (cnt.get(s[k]) ?? 0) + 1);

    // 统计值为 S[k] 的下标之和
    total.set(s[k], (total.get(s[k]) ?? 0) + k);
  }

  return res;
}
